tests/test_circulation_data.py

```diff
@@ -0,0 +1,265 @@
+    ContributorData,
+    FormatData,
+    IdentifierData,
+    LinkData,
+    ReplacementPolicy,
+    SubjectData,
+)
+
+from model import (
+    DataSource,
+    DeliveryMechanism,
+    Hyperlink, 
+    Identifier,
+    Representation,
+    RightsStatus,
+    Subject,
+)
+
+from . import (
+    DatabaseTest,
+    DummyHTTPClient,
+)
+
+from s3 import DummyS3Uploader
+
+class TestCirculationData(DatabaseTest):
+
+    def test_metadata_can_be_deepcopied(self):
+        # Check that we didn't put something in the CirculationData that
+        # will prevent it from being copied. (e.g., self.log)
+
+        subject = SubjectData(Subject.TAG, "subject")
+        contributor = ContributorData()
+        identifier = IdentifierData(Identifier.GUTENBERG_ID, "1")
+        link = LinkData(Hyperlink.OPEN_ACCESS_DOWNLOAD, "example.epub")
+        format = FormatData(Representation.EPUB_MEDIA_TYPE, DeliveryMechanism.NO_DRM)
+        rights_uri = RightsStatus.GENERIC_OPEN_ACCESS
+
+        circulation_data = CirculationData(
+            DataSource.GUTENBERG,
+            primary_identifier=identifier,
+            links=[link],
+            licenses_owned=5,
+            licenses_available=5,
+            licenses_reserved=None,
+            patrons_in_hold_queue=None,
+            formats=[format],
+            default_rights_uri=rights_uri,
+        )
+
+        circulation_data_copy = deepcopy(circulation_data)
+
+        # If deepcopy didn't throw an exception we're ok.
+        assert circulation_data_copy is not None
+
+
+    def test_links_filtered(self):
+        # Tests that passed-in links filter down to only the relevant ones.
+        link1 = LinkData(Hyperlink.OPEN_ACCESS_DOWNLOAD, "example.epub")
+        link2 = LinkData(rel=Hyperlink.IMAGE, href="http://example.com/")
+        link3 = LinkData(rel=Hyperlink.DESCRIPTION, content="foo")
+        link4 = LinkData(
+            rel=Hyperlink.THUMBNAIL_IMAGE, href="http://thumbnail.com/",
+            media_type=Representation.JPEG_MEDIA_TYPE,
+        )
+        link5 = LinkData(
+            rel=Hyperlink.IMAGE, href="http://example.com/", thumbnail=link4,
+            media_type=Representation.JPEG_MEDIA_TYPE,
+        )
+        links = [link1, link2, link3, link4, link5]
+
+        identifier = IdentifierData(Identifier.GUTENBERG_ID, "1")
+        circulation_data = CirculationData(
+            DataSource.GUTENBERG,
+            primary_identifier=identifier,
+            links=links,
+        )
+
+        filtered_links = sorted(circulation_data.links, key=lambda x:x.rel)
+
+        eq_([link1], filtered_links)
+
+
+    def test_open_access_content_mirrored(self):
+        # Make sure that open access material links are translated to our S3 buckets, and that 
+        # commercial material links are left as is.
+        # Note: Mirroring tests passing does not guarantee that all code now 
+        # correctly calls on CirculationData, as well as Metadata.  This is a risk.
```

Comment:
we don't require metadata to have a circulation, so it wouldn't be a complete guarantee.